Introduces a BreadcrumbsHelper

HavokInspiration opened the pull request:
This PR aims to introduce a `BreadcrumbsHelper` as discussed in #9542.
It exposes an API that allow precise insertions of crumbs (much like the way you'd manage your middlewares) to get around common issues related to the way Cake renders the views.

It also gives an easier way to customize the way the trail is rendered by using the `StringTemplateTrait`, the same way the `FormHelper` does.

The `render()` methods accepts two arguments : 
- the first argument will be used to customize the wrapper of the crumbs trail
- the second argument to customize the separator between crumbs. By default, no separator will be added between each crumbs.

For the templates, I went with what I think are sane default. Feel free to comment on this.

Once this is approved, I'll document the docblocks a bit more (to give more details about the arguments of the `render()` methods for instance and how to customize the "innerAttrs") and write something for the book.

Refs #9542

HavokInspiration committed:
Deprecate methods related to breadcrumbs in the HtmlHelper

HavokInspiration committed:
Implements the API to manage the crumbs trail

HavokInspiration committed:
Add the render method to the BreadcrumbsHelper

HavokInspiration committed:
Better customization option with the separator

HavokInspiration committed:
Fix PHP CS

ADmad commented:
Would be nice to have a method for adding multiple crumbs at once. It could be a separate method of 1st argument of `add()` could be overloaded to allow passing an array to it.

HavokInspiration commented:
Good idea. I'll do it in a separate methods.

ADmad commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,268 @@
+     */
+    protected function prepareLink($link)
+    {
+        if (is_string($link)) {
```
Why is this check necessary? `UrlHelper::build()` should be called for string URLs too, so that webroot can be properly appended to the URLs (especially for cases where app is in a subfolder).

HavokInspiration replied:
Good point.

ADmad commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,268 @@
+            'wrapper' => '<ul{{attrs}}>{{content}}</ul>',
+            'item' => '<li{{attrs}}><a href="{{link}}"{{innerAttrs}}>{{title}}</a></li>{{separator}}',
+            'itemWithoutLink' => '<li{{attrs}}><span{{innerAttrs}}>{{title}}</span></li>{{separator}}',
+            'separator' => '<li{{attrs}}><span{{innerAttrs}}>{{separator}}</span></li>'
```
It would be better to have separator as empty string by default IMO. Separators are generally handled with CSS. That's what popular frameworks like Bootstrap and Foundation do.

HavokInspiration replied:
True. But by default, separators are not displayed by the `render()` method.
A separator config has to be provided in the `render()` method (second argument) for the separator to be displayed. 
So using this with a framework should not be a problem if no separator config is provided.

dereuromark commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,268 @@
+                $found = true;
+                break;
+            }
+        }
```
you could extract the duplicate code in both insertBefore and insertAfter into a protected method.
This way it would also be easier to break early by just returning true.

HavokInspiration replied:
Yeah.

dereuromark commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,268 @@
+                $template = 'itemWithoutLink';
+            }
+
+            if (isset($separatorParams) && $key !== ($crumbsCount - 1)) {
```
It is cleaner to define it and just if ($var) check.

ADmad commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,268 @@
+                'title' => $title,
+                'link' => $link,
+                'separator' => '',
+            ];
```
`'templateVars' => isset($options['templateVars']) ? $options['templateVars'] : []` should be added to this array so that once can use custom template vars in the string template. Similarly options passed to all call to `formatTemplate()` should account for `templateVars`.

HavokInspiration committed:
Changes after review :
- code cleanup
- add the ability to use templateVars in all templates
- all urls are now set using the UrlHelper

HavokInspiration commented:
I just pushed a commit where I took your comments into account :

- code cleanup based on @dereuromark comments
- add the ability to use `templateVars` in all templates (not only items)
- all urls are now set using the `UrlHelper`

I did nothing about the separator comment you made @ADmad since in its current form, it will not cause trouble.
I'll try to add the ability to set multiple crumbs at once tonight if I get the time.

dereuromark commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,277 @@
+        $templater = $this->templater();
+
+        $separatorParams = [];
+        if (!empty($separator)) {
```
if ($separator) would be cleaner

HavokInspiration replied:
I'll never get used to this.

inoas replied:
I am not sure its a good idea for non-booleans either. But then `empty()` has its side effects, too.

dereuromark replied:
It is the same only without room for error.

dereuromark commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,277 @@
+     */
+    protected function prepareLink($link = null)
+    {
+        if (!empty($link)) {
```
if ($link)

HavokInspiration committed:
Code clean up

HavokInspiration committed:
Add a method to add multiple crumbs at once

- Also improve docblocks
- Code cleanup to follow guidelines

HavokInspiration commented:
I modified the `add()` method so it accepts its first argument to be an array to gives the ability to add multiple crumbs at once.
I also did the code cleanup to follow guidelines and updated the docblocks.

ADmad commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,318 @@
+    {
+        if (is_array($title)) {
+            foreach ($title as $crumb) {
+                $crumb = array_merge(['title' => '', 'link' => null, 'options' => []], $crumb);
```
These 2 lines can be merged into a single one 
```php
$this->crumbs[] = $crumb + ['title' => '', 'link' => null, 'options' => []];
```

ADmad commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,318 @@
+            return $this;
+        }
+
+        $this->crumbs[] = ['title' => $title, 'link' => $link, 'options' => $options];
```
Nitpick:  `$this->crumbs[] = compact('title, 'link', 'options');`  is shorter/cleaner :)  Similar change could be done in other functions too.

dereuromark replied:
The advantage of being explicit is that variables not defined are easier to spot. I would also be fine with keeping the current.

HavokInspiration replied:
I usually avoid using `compact()` for the reasons @dereuromark pointed out.

ADmad replied:
That's true in general but here all the vars are methods arguments so they are always going to be defined.

HavokInspiration replied:
Valid point.

ADmad commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,318 @@
+     * the link)
+     * - *templateVars*: Specific template vars in case you override the templates provided
+     * @return $this
+     */
```
`@throws` tag missing.

ADmad commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,318 @@
+     *
+     * @return null|string The URL of a crumb
+     */
+    protected function prepareLink($link = null)
```
This method seems unnecessary to me. The only usage of this method at line 243 can be written using ternary: 
```php
$link = $crumb['link'] ? $this->Url->build($crumb['link']) : null;
```
unless you foresee a use case where one might want to customize the URL generation.

HavokInspiration replied:
No we can remove it. If the need to customize URL generation is brought to the table, we can still restore it when needed, it would not cause too much work.

ADmad commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,318 @@
+     * - *templateVars*: Specific template vars in case you override the templates provided
+     * @return $this
+     */
+    public function add($title, $link = null, array $options = [])
```
Wouldn't argument name `$url` be more appropriate than `$link`?

ADmad committed:
Simplify code.

ADmad commented:
I pushed a commit to simplify the separator options handling in `render()` a bit. It was more convenient to do it myself rather than trying to explain :smile:

ADmad committed:
Avoid unnecessarily generating separator string multiple times.

dereuromark commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,312 @@
+            return $this;
+        }
+
+        $this->crumbs[] = ['title' => $title, 'link' => $link, 'options' => $options];
```
I do wonder however if the crumbs should be a simple Breadcrumb object with basic setters and getters instead of having to deal with arrays here, but maybe thats overhead. Even the crumb list could be a BreadcumbCollection in an OOP context.

HavokInspiration replied:
I would tend to agree if there was a lot of logic involved in each crumbs. At the moment, that seems unnecessary to me. Right now there are just a bunch of dumb texts elements.

HavokInspiration commented:
Comments taken into account. I will commit some more changes soon.

HavokInspiration committed:
Changes after review :
- remove the prepareLink method
- add missing throw tags
- code simplification / cleanup

HavokInspiration commented:
Changes pushed based on your comments @ADmad

HavokInspiration commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,295 @@
+    {
+        if (is_array($title)) {
+            foreach ($title as $crumb) {
+                $this->crumbs[] = $crumb + ['title' => '', 'url' => null, 'options' => []];
```
I agree about ``prepend`` and ``insertAt``, but not for this line.

HavokInspiration replied:
I agree on the principle @dereuromark and usually avoid using `compact()` except in some specific cases, but impact here is really minimal and as pointed out before, it is only used on variables that are method arguments with default.

ADmad replied:
@HavokInspiration My bad, the values here are defaults not the arguments, so this is fine as it is.

HavokInspiration committed:
Use compact instead of defining arrays

markstory commented on tests/TestCase/View/Helper/BreadcrumbsHelperTest.php:
```diff
@@ -0,0 +1,424 @@
+
+        $result = $this->breadcrumbs->render(
+            ['data-stuff' => 'foo and bar'],
+            ['separator' => '<i class="fa fa-angle-right"></i>', 'class' => 'separator', 'templateVars' => ['custom' => 'custom']]
```
How does `custom` end up in the resulting HTML? I don't see where the templates are changed.

HavokInspiration replied:
Sharp eye.
It actually comes from the templates define in the Helper itself. Which is bad. Good catch, I'll update the PR.

HavokInspiration replied:
Done

HavokInspiration committed:
Fix a custom variable from the separator template and fix the tests

markstory reviewed (approved):
Looks good to me.

ravage84 commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,295 @@
+    /**
+     * Add a crumb to the trail.
+     *
+     * @param string|array $title Title of the crumb. If provided as an array, it will add each values of the array
```
Instead of "If provided as an array...", I would start with "Alternatively, an array with..." or something like that.
At first it sounds like you can give the single title as array.

ravage84 commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,295 @@
+     */
+    public function add($title, $url = null, array $options = [])
+    {
+        if (is_array($title)) {
```
Looking at the code and the doc blocks, wouldn't it be cleaner to make the multiple breadcrumbs methods separate methods?

ravage84 commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,295 @@
+    /**
+     * Prepend a crumb to the start of the queue.
+     *
+     * @param string $title Title of the crumb
```
So here we cannot prepend multiple breadcrumbs as in ``add()``?

ravage84 commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,295 @@
+     * - *templateVars*: Specific template vars in case you override the templates provided
+     * @return $this
+     */
+    public function add($title, $url = null, array $options = [])
```
May be call it ``append()`` instead, so it's crystal clear what it does?

ravage84 commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,295 @@
+    {
+        $key = $this->findCrumb($matchingTitle);
+
+        if ($key !== null) {
```
I would write it this way:

```` php 
$key = $this->findCrumb($matchingTitle);
if ($key === null) {
    throw new LogicException(sprintf("No crumb matching '%s' could be found.", $matchingTitle));
}

return $this->insertAt($key, $title, $url, $options);
````

HavokInspiration replied:
Looks better indeed.

ravage84 commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,295 @@
+    {
+        $key = $this->findCrumb($matchingTitle);
+
+        if ($key !== null) {
```
Same as with ``insertBefore()``.

ravage84 commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,295 @@
+    }
+
+    /**
+     * Returns the crumbs list
```
"crumb list"?

ravage84 commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,295 @@
+    ];
+
+    /**
+     * The crumbs list.
```
"The crumb list", I believe. But I'm not a English native speaker.

ravage84 commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,295 @@
+     * and the existing element will be shifted one index greater.
+     *
+     * @param int $index The index to insert at.
+     * @param string $title Title of the crumb
```
General nitpicking: Sometimes you add dots at the end of annotations, sometimes you do not. I think we generally settled with dots at the end.

HavokInspiration replied:
Do we have a standard settled for this ?
It would good to have a sniff for that if so. This is typically the kind of thing I absolutely don't pay attention to.

ADmad replied:
> It would good to have a sniff for that if so.

We did have a [sniff](https://github.com/cakephp/cakephp-codesniffer/blob/master/CakePHP/ruleset.xml#L28) for this but it has been disabled since existing codebase has lot of such inconsistencies and no one has the time to actually fix them.

ravage84 commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,295 @@
+    /**
+     * Insert a crumb before the first matching crumb with the specified title.
+     *
+     * Finds the index of the first middleware that matches the provided class,
```
middleware? I think that's a copy pasta error.

ravage84 commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,295 @@
+    /**
+     * Insert a crumb after the first matching crumb with the specified title.
+     *
+     * Finds the index of the first middleware that matches the provided class,
```
middleware!?

HavokInspiration replied:
Yeah... most of the managing code is from the MiddlewareQueue.

ravage84 commented on src/View/Helper/BreadcrumbsHelper.php:
```diff
@@ -0,0 +1,295 @@
+        }
+
+        $crumbTrail = '';
+        foreach ($crumbs as $key => $crumb) {
```
May be this foreach could even be extracted as protected method. Could make subclassing/extending easier. Not a big deal though.

ravage84 reviewed (changes requested):
Great work @HavokInspiration!

dereuromark commented on tests/TestCase/View/Helper/BreadcrumbsHelperTest.php:
```diff
@@ -0,0 +1,426 @@
+
+    /**
+     * Tests the render method with template vars
+     * @return void
```
In the first comment of this Kind you could have said "and others" instead.

HavokInspiration replied:
We should have a sniff for that (or do we already have one ?) if it is a mandatory thing.
@dereuromark since you seem to be the sniff specialist, do you think that is something that can be achieved ?

ADmad replied:
@HavokInspiration As long as phpcs is happy you can ignore any docblock related issues in tests.

dereuromark replied:
I already have such a sniff that can be ported to Cake easily.

HavokInspiration commented:
I would like to have opinions from others about the following points raised by @ravage84 : 
- having a dedicated methods to add multiple crumbs at once (instead of just using the same method)
- changing the name of the `add` method to `append` (and thus, losing consistency with the MiddlewareQueue from which I got my inspiration)
- throw an exception in `insertAt` if the index is out of bound
- have the prepend method (and possibly all other methods ?) allowing adding multiple crumbs at once as well

HavokInspiration committed:
Fix docblocks and rearrange some codes

HavokInspiration committed:
Throws an exception if the index is out of bound in `insertAt()`

HavokInspiration committed:
Update the tests for the exception of `insertAt()`

HavokInspiration committed:
The prepend method can now add multiple crumbs at once (in the order given)

HavokInspiration committed:
More docblocks fix

HavokInspiration commented:
Prepend now accepts bulk insertions
`insertAt()` throws an exception if the index is out of bounds
Doc block fixed

ChristopherCastro commented:
I saw this PR and I would like to mention that, usually, breadcrumbs are queued out of the View-layer scope. In some complex applications different parts of the application itself (Plugins, Cells, etc) may require to insert elements into the queue.

I found myself in a similar situation time ago, so I made a Breadcrumb helper, but using a static approach based on a "CrumbRegistry" class, the idea is that anyone is allowed to insert crumbs into the queue at any point of the script exceution, and the Helper is used for the rendering purposes mainly.

just an opinion

ADmad commented:
> In some complex applications different parts of the application itself (Plugins, Cells, etc) may require to insert elements into the queue.

You can maintain the list of items/links in your own datastore which suits you needs and then pass the whole list to `BreadcrumbsHelper::add()` for rendering.

ChristopherCastro commented:
> You can maintain the list of items/links in your own datastore which suits you needs and then pass the whole list to BreadcrumbsHelper::add() for rendering.

The keyword here is "pass", and that's the main problem I'm trying to describe, as it would require (probably) a hard coupling between Controler and Views. Passing variables from here to there, however a static approach would allow us to avoid passing things and access the crumbs queue from anywhere.

dereuromark commented:
Looks like we are ready to merge.
